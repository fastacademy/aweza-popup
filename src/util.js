function base64ToArrayBuffer (base64) {
  const binary_string = window.atob(base64)
  const len = binary_string.length
  const bytes = new Uint8Array(len)
  for (let i = 0; i < len; i++) {
    bytes[i] = binary_string.charCodeAt(i)
  }
  return bytes.buffer
}

export function createAudioBlobUrl (base64Data) {
  const blob = new Blob([base64ToArrayBuffer(base64Data)], {type: 'audio/mp3'})
  return window.URL.createObjectURL(blob)
}
