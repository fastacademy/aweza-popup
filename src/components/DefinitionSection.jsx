import { h, Component } from 'preact'
import { createAudioBlobUrl } from '../util'
import Speaker from './Speaker'

export default class DefinitionSection extends Component {
  render () {
    return (
      <div class="aweza-definition">
        <div class="aweza-section-title">
          <b>Definition</b>
        </div>
        <div class="aweza-section-content">
          <DefinitionText {...this.props.term}/>
          <DefinitionDescription {...this.props.term}/>
          {this.props.translationTerm &&
          [
            <hr/>,
            <DefinitionText {...this.props.translationTerm}/>,
            <DefinitionDescription {...this.props.translationTerm}/>]
          }
        </div>
      </div>
    )
  }
}

class DefinitionText extends Component {
  render () {
    let audioUrl = null
    if (this.props.audio) {
      audioUrl = this.props.audio
    }
    else if (this.props.tts.text) {
      audioUrl = createAudioBlobUrl(this.props.tts.text)
    }
    return (
      <div class="aweza-definition-text aweza-section-content-text">
        <div class="aweza-definition-text-left">
          <Speaker uniqueId={`aweza-term-${this.props.id}-text-audio`} audioUrl={audioUrl}/>
          <div class="aweza-definition-content-text-text">
            <div style={{fontSize: 15}}>{this.props.text}</div>
          </div>
        </div>
        <div class="aweza-definition-content-text-right language-code">
          {this.props.language.code.toUpperCase()}
        </div>
      </div>
    )
  }
}

class DefinitionDescription extends Component {
  render () {
    let audioUrl = null
    if (this.props.description_audio) {
      audioUrl = this.props.description_audio
    }
    else if (this.props.tts.description) {
      audioUrl = createAudioBlobUrl(this.props.tts.description)
    }
    return (
      <div class="aweza-definition-description aweza-section-content-text">
        <Speaker uniqueId={`aweza-term-${this.props.id}-definition-audio`} audioUrl={audioUrl}/>
        <div class="aweza-definition-description">
          <div dangerouslySetInnerHTML={{__html: this.props.description}}/>
        </div>
      </div>
    )
  }
}