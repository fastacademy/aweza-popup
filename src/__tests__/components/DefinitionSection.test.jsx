// noinspection ES6UnusedImports
import { h } from 'preact';
import render from 'preact-render-to-json';
import {terms} from '../dummydata';
import DefinitionSection from "../../components/DefinitionSection";
import {createAudioBlobUrl} from "../../util";

function getComponent () {
    return (
        <DefinitionSection
            term={terms[0]}
            translationTerm={terms[0].translations[0]}
        />
    );
}

test('DefinitionSection component', () => {
    global.URL.createObjectURL = jest.fn();
    const tree = render(getComponent());

    expect(tree).toMatchSnapshot();
});
