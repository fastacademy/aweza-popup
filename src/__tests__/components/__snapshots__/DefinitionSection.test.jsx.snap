// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefinitionSection component 1`] = `
<div
  class="aweza-definition"
>
  <div
    class="aweza-section-title"
  >
    <b>
      Definition
    </b>
  </div>
  <div
    class="aweza-section-content"
  >
    <div
      class="aweza-definition-text aweza-section-content-text"
    >
      <div
        class="aweza-definition-text-left"
      >
        <div
          class="aweza-definition-content-text-text"
        >
          <div
            style={
              Object {
                "fontSize": 15,
              }
            }
          >
            IParabola
          </div>
        </div>
      </div>
      <div
        class="aweza-definition-content-text-right language-code"
      >
        XH
      </div>
    </div>
    <div
      class="aweza-definition-description aweza-section-content-text"
    >
      <div
        class="aweza-definition-description"
      >
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Igophe elijonge phezulu okanye ezantsi, eliqala kwipoyinti c kwindawo ka-y lize liqhubeke ngokungapheliyo.</p>",
            }
          }
        />
      </div>
    </div>
    <hr />
    <div
      class="aweza-definition-text aweza-section-content-text"
    >
      <div
        class="aweza-definition-text-left"
      >
        <div
          class="aweza-definition-content-text-text"
        >
          <div
            style={
              Object {
                "fontSize": 15,
              }
            }
          >
            Parabola
          </div>
        </div>
      </div>
      <div
        class="aweza-definition-content-text-right language-code"
      >
        EN
      </div>
    </div>
    <div
      class="aweza-definition-description aweza-section-content-text"
    >
      <div
        class="aweza-definition-description"
      >
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "An up or down facing curve, starting at a point c on the y-axis and continuing to infinity.",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
